File: doc/release.pl
#!/usr/bin/perl
####################################################################################################################################
# release.pl - PgBackRest Release Manager
####################################################################################################################################

####################################################################################################################################
# Perl includes
####################################################################################################################################
use strict;
use warnings FATAL => qw(all);
use Carp qw(confess);
use English '-no_match_vars';

$SIG{__DIE__} = sub { Carp::confess @_ };

use Cwd qw(abs_path);
use File::Basename qw(dirname);
use Getopt::Long qw(GetOptions);
use Pod::Usage qw(pod2usage);
use Storable;

use lib dirname($0) . '/lib';
use lib dirname(dirname($0)) . '/build/lib';
use lib dirname(dirname($0)) . '/lib';
use lib dirname(dirname($0)) . '/test/lib';

use pgBackRestTest::Common::ExecuteTest;
use pgBackRestTest::Common::Storage;
use pgBackRestTest::Common::StoragePosix;
use pgBackRestTest::Common::VmTest;

use pgBackRestDoc::Common::Doc;
use pgBackRestDoc::Common::DocConfig;
use pgBackRestDoc::Common::DocManifest;
use pgBackRestDoc::Common::DocRender;
use pgBackRestDoc::Common::Exception;
use pgBackRestDoc::Common::Log;
use pgBackRestDoc::Common::String;
use pgBackRestDoc::Custom::DocCustomRelease;
use pgBackRestDoc::Html::DocHtmlSite;
use pgBackRestDoc::Latex::DocLatex;
use pgBackRestDoc::Markdown::DocMarkdown;
use pgBackRestDoc::ProjectInfo;

####################################################################################################################################
# Usage
####################################################################################################################################

=head1 NAME

release.pl - pgBackRest Release Manager

=head1 SYNOPSIS

release.pl [options]

 General Options:
   --help           Display usage and exit
   --version        Display pgBackRest version
   --quiet          Sets log level to ERROR
   --log-level      Log level for execution (e.g. ERROR, WARN, INFO, DEBUG)

 Release Options:
   --build          Build the cache before release (should be included in the release commit)
   --deploy         Deploy documentation to website (can be done as docs are updated)
   --no-gen         Don't auto-generate
   --vm             vm to build documentation for
=cut

####################################################################################################################################
# Load command line parameters and config (see usage above for details)
####################################################################################################################################
my $bHelp = false;
my $bVersion = false;
my $bQuiet = false;
my $strLogLevel = 'info';
my $bBuild = false;
my $bDeploy = false;
my $bNoGen = false;
my $strVm = undef;

GetOptions ('help' => \$bHelp,
            'version' => \$bVersion,
            'quiet' => \$bQuiet,
            'log-level=s' => \$strLogLevel,
            'build' => \$bBuild,
            'deploy' => \$bDeploy,
            'no-gen' => \$bNoGen,
            'vm=s' => \$strVm)
    or pod2usage(2);

####################################################################################################################################
# Run in eval block to catch errors
####################################################################################################################################
eval
{
    # Display version and exit if requested
    if ($bHelp || $bVersion)
    {
        print PROJECT_NAME . ' ' . PROJECT_VERSION . " Release Manager\n";

        if ($bHelp)
        {
            print "\n";
            pod2usage();
        }

        exit 0;
    }

    # If neither build nor deploy is requested then error
    if (!$bBuild && !$bDeploy)
    {
        confess &log(ERROR, 'neither --build nor --deploy requested, nothing to do');
    }

    # Set console log level
    if ($bQuiet)
    {
        $strLogLevel = 'error';
    }

    logLevelSet(undef, uc($strLogLevel), OFF);

    # Set the paths
    my $strDocPath = dirname(abs_path($0));
    my $strDocHtml = "${strDocPath}/output/html";
    my $strDocExe = "${strDocPath}/doc.pl";
    my $strTestExe = dirname($strDocPath) . "/test/test.pl";

    my $oStorageDoc = new pgBackRestTest::Common::Storage(
        $strDocPath, new pgBackRestTest::Common::StoragePosix({bFileSync => false, bPathSync => false}));

    # Determine if this is a dev release
    my $bDev = PROJECT_VERSION =~ /dev$/;
    my $strVersion = $bDev ? 'dev' : PROJECT_VERSION;

    # Make sure version number matches the latest release
    &log(INFO, "check version info");

    my $strReleaseFile = dirname(dirname(abs_path($0))) . '/doc/xml/release.xml';
    my $oRelease = (new pgBackRestDoc::Custom::DocCustomRelease(new pgBackRestDoc::Common::Doc($strReleaseFile)))->releaseLast();

    if ($oRelease->paramGet('version') ne PROJECT_VERSION)
    {
        confess 'unable to find version ' . PROJECT_VERSION . " as the most recent release in ${strReleaseFile}";
    }

    if ($bBuild)
    {
        if (!$bNoGen)
        {
            # Update git history
            my $strGitCommand =
                'git -C ' . $strDocPath .
                ' log --pretty=format:\'{^^^^commit^^^^:^^^^%H^^^^,^^^^date^^^^:^^^^%ci^^^^,^^^^subject^^^^:^^^^%s^^^^,^^^^body^^^^:^^^^%b^^^^},\'';
            my $strGitLog = qx($strGitCommand);
            $strGitLog =~ s/\^\^\^\^\}\,\n/\#\#\#\#/mg;
            $strGitLog =~ s/\\/\\\\/g;
            $strGitLog =~ s/\n/\\n/mg;
            $strGitLog =~ s/\r/\\r/mg;
            $strGitLog =~ s/\t/\\t/mg;
            $strGitLog =~ s/\"/\\\"/g;
            $strGitLog =~ s/\^\^\^\^/\"/g;
            $strGitLog =~ s/\#\#\#\#/\"\}\,\n/mg;
            $strGitLog = '[' . substr($strGitLog, 0, length($strGitLog) - 1) . ']';
            my @hyGitLog = @{(JSON::PP->new()->allow_nonref())->decode($strGitLog)};

            # Load prior history
            my @hyGitLogPrior = @{(JSON::PP->new()->allow_nonref())->decode(
                ${$oStorageDoc->get("${strDocPath}/resource/git-history.cache")})};

            # Add new commits
            for (my $iGitLogIdx = @hyGitLog - 1; $iGitLogIdx >= 0; $iGitLogIdx--)
            {
                my $rhGitLog = $hyGitLog[$iGitLogIdx];
                my $bFound = false;

                foreach my $rhGitLogPrior (@hyGitLogPrior)
                {
                    if ($rhGitLog->{commit} eq $rhGitLogPrior->{commit})
                    {
                        $bFound = true;
                    }
                }

                next if $bFound;

                $rhGitLog->{body} = trim($rhGitLog->{body});

                if ($rhGitLog->{body} eq '')
                {
                    delete($rhGitLog->{body});
                }

                unshift(@hyGitLogPrior, $rhGitLog);
            }

            # Write git log
            $strGitLog = undef;

            foreach my $rhGitLog (@hyGitLogPrior)
            {
                $strGitLog .=
                    (defined($strGitLog) ? ",\n" : '') .
                    "    {\n" .
                    '        "commit": ' . trim((JSON::PP->new()->allow_nonref()->pretty())->encode($rhGitLog->{commit})) . ",\n" .
                    '        "date": ' . trim((JSON::PP->new()->allow_nonref()->pretty())->encode($rhGitLog->{date})) . ",\n" .
                    '        "subject": ' . trim((JSON::PP->new()->allow_nonref()->pretty())->encode($rhGitLog->{subject}));

                # Skip the body if it is empty or a release (since we already have the release note content)
                if ($rhGitLog->{subject} !~ /^v[0-9]{1,2}\.[0-9]{1,2}\: /g && defined($rhGitLog->{body}))
                {
                    $strGitLog .=
                        ",\n" .
                        '        "body": ' . trim((JSON::PP->new()->allow_nonref()->pretty())->encode($rhGitLog->{body}));
                }

                $strGitLog .=
                    "\n" .
                    "    }";
            }

            $oStorageDoc->put("${strDocPath}/resource/git-history.cache", "[\n${strGitLog}\n]\n");

            # Generate coverage summary
            &log(INFO, "Generate Coverage Summary");
            executeTest("${strTestExe} --vm=u22 --no-valgrind --clean --coverage-summary", {bShowOutputAsync => true});
        }

        # Remove permanent cache file
        $oStorageDoc->remove("${strDocPath}/resource/exe.cache", {bIgnoreMissing => true});

        # Remove all docker containers to get consistent IP address assignments
        executeTest('docker rm -f $(docker ps -a -q)', {bSuppressError => true});

        # Generate deployment docs for RHEL
        if (!defined($strVm) || $strVm eq VM_RH8)
        {
            &log(INFO, "Generate RHEL documentation");

            executeTest("${strDocExe} --deploy --key-var=os-type=rhel --out=pdf", {bShowOutputAsync => true});

            if (!defined($strVm))
            {
                executeTest("${strDocExe} --deploy --cache-only --key-var=os-type=rhel --out=pdf");
            }
        }

        # Generate deployment docs for Debian
        if (!defined($strVm) || $strVm eq VM_U18)
        {
            &log(INFO, "Generate Debian/Ubuntu documentation");

            executeTest("${strDocExe} --deploy --out=man --out=html --out=markdown", {bShowOutputAsync => true});
        }

        # Generate a full copy of the docs for review
        if (!defined($strVm))
        {
            &log(INFO, "Generate full documentation for review");

            executeTest(
                "${strDocExe} --deploy --out-preserve --cache-only --key-var=os-type=rhel --out=html" .
                    " --var=project-url-root=index.html");
            $oStorageDoc->move("$strDocHtml/user-guide.html", "$strDocHtml/user-guide-rhel.html");

            executeTest(
                "${strDocExe} --deploy --out-preserve --cache-only --out=man --out=html --var=project-url-root=index.html");
        }
    }

    if ($bDeploy)
    {
        my $strDeployPath = "${strDocPath}/site";

        # Generate docs for the website history
        &log(INFO, 'Generate website ' . ($bDev ? 'dev' : 'history') . ' documentation');

        my $strDocExeVersion =
            ${strDocExe} . ($bDev ? ' --dev' : ' --deploy --cache-only') . ' --var=project-url-root=index.html --out=html';

        executeTest("${strDocExeVersion} --out-preserve --key-var=os-type=rhel");
        $oStorageDoc->move("$strDocHtml/user-guide.html", "$strDocHtml/user-guide-rhel.html");

        $oStorageDoc->remove("$strDocHtml/release.html");
        executeTest("${strDocExeVersion} --out-preserve --exclude=release");

        # Deploy to repository
        &log(INFO, '...Deploy to repository');
        executeTest("rm -rf ${strDeployPath}/prior/${strVersion}");
        executeTest("mkdir ${strDeployPath}/prior/${strVersion}");
        executeTest("cp ${strDocHtml}/* ${strDeployPath}/prior/${strVersion}");

        # Generate docs for the main website
        if (!$bDev)
        {
            &log(INFO, "Generate website documentation");

            executeTest("${strDocExe} --var=analytics=y --deploy --cache-only --key-var=os-type=rhel --out=html");
            $oStorageDoc->move("$strDocHtml/user-guide.html", "$strDocHtml/user-guide-rhel.html");
            executeTest("${strDocExe} --var=analytics=y --deploy --out-preserve --cache-only --out=html");

            # Deploy to repository
            &log(INFO, '...Deploy to repository');
            executeTest("rm -rf ${strDeployPath}/dev");
            executeTest("find ${strDeployPath} -maxdepth 1 -type f -exec rm {} +");
            executeTest("cp ${strDocHtml}/* ${strDeployPath}");
            executeTest("cp ${strDocPath}/../README.md ${strDeployPath}");
            executeTest("cp ${strDocPath}/../LICENSE ${strDeployPath}");
        }

        # Update permissions
        executeTest("find ${strDeployPath} -type d -exec chmod 750 {} +");
        executeTest("find ${strDeployPath} -type f -exec chmod 640 {} +");
    }

    # Exit with success
    exit 0;
}

####################################################################################################################################
# Check for errors
####################################################################################################################################
or do
{
    # If a backrest exception then return the code
    exit $EVAL_ERROR->code() if (isException(\$EVAL_ERROR));

    # Else output the unhandled error
    print $EVAL_ERROR;
    exit ERROR_UNHANDLED;
};

# It shouldn't be possible to get here
&log(ASSERT, 'execution reached invalid location in ' . __FILE__ . ', line ' . __LINE__);
exit ERROR_ASSERT;

File: doc/release.py
#!/usr/bin/env python3
"""pgBackRest Release Manager.

Builds the documentation for a release, deploys it to the website, and builds the copy that ships in the distribution. A separate
program from the documentation builder because it is about a release rather than about a document: it decides which documentation to
build, for which platform, and where it goes.

The documentation is built for each platform it documents, since the commands a reader runs differ between them, and the copy of the
user guide for each is kept under a name of its own.

All output, including errors, goes to stdout so the run reads in the order it happened."""

####################################################################################################################################
import argparse
import json
import os
import re
import shutil
import signal
import sys

# Send everything written to stderr to stdout instead so the output is in the order it happened
sys.stderr = sys.stdout

# Do not cache bytecode. The tool runs from the source tree, where a __pycache__ would show up as an unexpected binary file in the
# linter and in the distribution. This must be set before the library modules are imported below.
sys.dont_write_bytecode = True

# Each tool keeps its library beside itself and may use the libraries below it in the hierarchy. Insert them first, lowest last, so
# the doc modules are found before anything else on the path.
for lib in ("build", "doc"):
    sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), lib, "lib"))

from command.render.release import release_last  # noqa: E402
from common.error import EXIT_ERROR, EXIT_TERM, ToolError, error_trace  # noqa: E402
from common.exec import exec_result  # noqa: E402
from common.log import *  # noqa: E402
from common.storage import file_read, file_remove, file_write  # noqa: E402
from common.xml import xml_document_parse, xml_node_attribute, xml_node_normalize  # noqa: E402
from config.project import PROJECT_NAME, project_version  # noqa: E402

# Where the documentation is built and where the release keeps what it needs
_FILE_HISTORY = "resource/git-history.cache"
_FILE_CACHE = "resource/exe.cache"
_PATH_HTML = "output/html"
_PATH_SITE = "site"

# Files at the top of the site that belong to the site rather than to a build of the documentation, so a deploy leaves them alone.
# CNAME is what points the domain at the site, so removing it takes the site off its domain until someone puts it back.
_SITE_KEEP_LIST = ("CNAME",)

# The distributions the documentation is built for, named for the os type the documentation generates commands for. The user guide
# for each is kept under a name of its own, since a reader picks the one for the system they are on.
_DISTRO_LIST = ("debian", "rhel")

_FILE_USER_GUIDE = "user-guide.html"
_FILE_USER_GUIDE_RHEL = "user-guide-rhel.html"

# What the release notes call a release, which is the subject of the commit that closed it. The body of such a commit is the release
# notes themselves, so it is not kept in the history as well.
_RELEASE_SUBJECT_EXP = re.compile(r"^v[0-9]{1,2}\.[0-9]{1,2}(\.[0-9]+)?: ")

# Fields of a commit and the order they are written in, which is the order a reader would want them
_HISTORY_FIELD = ("commit", "date", "subject", "body")


####################################################################################################################################
def cfg_load(arg_list, path_repo):
    """Parse the command line and apply the rules that cannot be expressed in the parser."""

    parser = argparse.ArgumentParser(prog="release.py", description="pgBackRest Release Manager")

    parser.add_argument("--version", action="version", version="pgBackRest %s Release Manager" % project_version(path_repo))
    parser.add_argument("--build", action="store_true", help="build the cache before release (include in the release commit)")
    parser.add_argument("--deploy", action="store_true", help="deploy the documentation to the website")
    parser.add_argument("--dist", action="store_true", help="build the documentation that ships in the distribution")
    parser.add_argument("--no-gen", dest="gen", action="store_false", help="do not regenerate the git history and coverage summary")
    parser.add_argument("--no-exe", dest="exe", action="store_false", help="do not run the commands (only applies to --dist)")
    parser.add_argument(
        "--distro", choices=_DISTRO_LIST, help="build the documentation for one distribution rather than for every one"
    )
    parser.add_argument("--quiet", action="store_true", help="set the log level to error")
    parser.add_argument(
        "--log-level", default="info", choices=sorted(LEVEL_NAME.values()), metavar="LEVEL", help="console log level"
    )

    config = parser.parse_args(arg_list)
    config.repo_path = path_repo
    config.doc_path = os.path.join(path_repo, "doc")

    if not (config.build or config.deploy or config.dist):
        raise ToolError("neither --build nor --deploy nor --dist requested, nothing to do")

    config.log_level = OFF if config.quiet else log_level_parse(config.log_level)

    return config


####################################################################################################################################
def doc_build(config, arg_list, show_output=False):
    """Build the documentation."""

    exec_result(
        "%s %s" % (os.path.join(config.doc_path, "doc.py"), " ".join(arg_list)),
        suppress_stderr=True,
        show_output=show_output,
    )


####################################################################################################################################
def user_guide_rename(config):
    """Keep the user guide that was just built under a name of its own, so the next one does not replace it."""

    path_html = os.path.join(config.doc_path, _PATH_HTML)

    os.replace(os.path.join(path_html, _FILE_USER_GUIDE), os.path.join(path_html, _FILE_USER_GUIDE_RHEL))


####################################################################################################################################
def history_update(config):
    """Add what has happened since the last release to the history the release notes are checked against.

    The history is kept in the repository because it is fixed once a release is out and reading it from git on every documentation
    build is slow. A commit that closed a release does not keep its body, since the release notes already say what it said."""

    log(INFO, "update git history")

    # Records are separated by a byte that cannot appear in a commit message, and so are the fields within a record
    output = exec_result(
        "git -C %s log --pretty=format:%%H%%x00%%ci%%x00%%s%%x00%%b%%x1e" % config.repo_path, suppress_stderr=True
    )[1]

    history = []

    for record in output.split("\x1e"):
        record = record.strip("\n")

        if record == "":
            continue

        commit, date, subject, body = record.split("\x00", 3)
        entry = {"commit": commit, "date": date, "subject": subject}
        body = body.strip()

        if body != "":
            entry["body"] = body

        history.append(entry)

    path_history = os.path.join(config.doc_path, _FILE_HISTORY)
    result = json.loads(file_read(path_history))
    seen = {entry["commit"] for entry in result}

    # Walk what git reported oldest first so that adding each to the front leaves the newest first
    for entry in reversed(history):
        if entry["commit"] not in seen:
            result.insert(0, entry)

    file_write(
        path_history,
        json.dumps(
            [
                {
                    field: entry[field]
                    for field in _HISTORY_FIELD
                    if field in entry and not (field == "body" and _RELEASE_SUBJECT_EXP.match(entry["subject"]))
                }
                for entry in result
            ],
            indent=4,
            ensure_ascii=False,
        )
        + "\n",
    )


####################################################################################################################################
def cmd_build(config):
    """Build the documentation for a release, along with the cache that lets it be built again without running anything."""

    if config.gen:
        history_update(config)

        log(INFO, "generate coverage summary")

        exec_result(
            "%s --vm=u24 --no-valgrind --clean --coverage-summary" % os.path.join(config.repo_path, "test/test.py"),
            suppress_stderr=True,
            show_output=True,
        )

    # The cache is built from nothing, since what it holds is what this release does rather than what the last one did
    file_remove(os.path.join(config.doc_path, _FILE_CACHE))

    # Remove the hosts of any previous build so the addresses they are given are the same every time
    host_remove()

    if config.distro is None or config.distro == "rhel":
        log(INFO, "generate RHEL documentation")

        doc_build(config, ["--deploy", "--key-var=os-type=rhel", "--out=html"], show_output=True)

    if config.distro is None or config.distro == "debian":
        log(INFO, "generate Debian/Ubuntu documentation")

        doc_build(config, ["--deploy", "--out=man", "--out=html", "--out=markdown"], show_output=True)

    # A copy of everything for review, which is the documentation as the website will have it
    if config.distro is None:
        log(INFO, "generate full documentation for review")

        doc_build(
            config,
            [
                "--deploy",
                "--out-preserve",
                "--cache-only",
                "--key-var=os-type=rhel",
                "--out=html",
                "--var=project-url-root=index.html",
            ],
        )
        user_guide_rename(config)

        doc_build(
            config, ["--deploy", "--out-preserve", "--cache-only", "--out=man", "--out=html", "--var=project-url-root=index.html"]
        )


####################################################################################################################################
def host_remove():
    """Remove the hosts of a documentation build.

    The addresses hosts are given depend on what else is running, and the documentation shows those addresses, so a build starts
    from nothing."""

    name_list = exec_result("docker ps -a --format '{{.Names}}'", suppress_stderr=True)[1].split()
    doc_list = [name for name in name_list if name.startswith("doc-")]

    if len(doc_list) > 0:
        exec_result("docker rm -f %s" % " ".join(doc_list), suppress_error=True)


####################################################################################################################################
def cmd_deploy(config, dev):
    """Deploy the documentation to the website."""

    path_html = os.path.join(config.doc_path, _PATH_HTML)
    path_site = os.path.join(config.doc_path, _PATH_SITE)
    version = "dev" if dev else project_version(config.repo_path)

    log(INFO, "generate website %s documentation" % ("dev" if dev else "history"))

    # A development build is rendered as it is, and a release is rendered from the cache so it comes out the same every time
    arg_list = (["--dev"] if dev else ["--deploy", "--cache-only"]) + ["--var=project-url-root=index.html", "--out=html"]

    doc_build(config, arg_list + ["--out-preserve", "--key-var=os-type=rhel"])
    user_guide_rename(config)

    file_remove(os.path.join(path_html, "release.html"))
    doc_build(config, arg_list + ["--out-preserve", "--exclude=release"])

    log(INFO, "...deploy to repository")

    path_prior = os.path.join(path_site, "prior", version)

    shutil.rmtree(path_prior, ignore_errors=True)
    shutil.copytree(path_html, path_prior)

    # The main website is only replaced by a release, since a development build is of what is not out yet
    if not dev:
        log(INFO, "generate website documentation")

        doc_build(config, ["--var=analytics=y", "--deploy", "--cache-only", "--key-var=os-type=rhel", "--out=html"])
        user_guide_rename(config)
        doc_build(config, ["--var=analytics=y", "--deploy", "--out-preserve", "--cache-only", "--out=html"])

        log(INFO, "...deploy to repository")

        shutil.rmtree(os.path.join(path_site, "dev"), ignore_errors=True)

        for name in os.listdir(path_site):
            path = os.path.join(path_site, name)

            # Only what this build writes, which is regular files the site does not own. A link is the site's too, e.g. a retired
            # platform name pointed at the guide that replaced it, so it is left where it is.
            if name not in _SITE_KEEP_LIST and os.path.isfile(path) and not os.path.islink(path):
                os.remove(path)

        shutil.copytree(path_html, path_site, dirs_exist_ok=True)

        for name in ("README.md", "LICENSE"):
            shutil.copyfile(os.path.join(config.repo_path, name), os.path.join(path_site, name))

    exec_result("find %s -path .git -prune -type d -exec chmod 750 {} +" % path_site)
    exec_result("find %s -path .git -prune -type f -exec chmod 640 {} +" % path_site)


####################################################################################################################################
def cmd_dist(config, dev):
    """Build the documentation that ships in the distribution.

    What ships is the documentation of the project rather than of the website, so what is only of interest on the website is left
    out."""

    log(INFO, "generate dist documentation")

    arg_list = [
        "--var=project-url-root=index.html",
        "--exclude=metric",
        "--exclude=news",
        "--var=card=n",
        "--var=news=n",
        "--var=sponsor=n",
    ]

    if not config.exe:
        arg_list.append("--no-exe")
    elif not dev:
        arg_list += ["--cache-only", "--var=release-date-static=y"]

    doc_build(config, arg_list + ["--key-var=os-type=rhel", "--out=html"])
    user_guide_rename(config)
    doc_build(config, arg_list + ["--out-preserve", "--out=html", "--out=man"])

    # The readme the distribution tarball ships is generated rather than committed, so it is rendered on its own
    doc_build(config, ["--out=markdown", "--include=distribution"])


####################################################################################################################################
def command_run(config):
    """Do what was asked for."""

    version = project_version(config.repo_path)
    dev = version.endswith("dev")

    # The version the code says it is must be the version the release notes say is most recent, since the documentation reports it
    log(INFO, "check version info")

    path_release = os.path.join(config.doc_path, "xml/release.xml")
    release = xml_document_parse(file_read(path_release), path_release)

    xml_node_normalize(release)

    if xml_node_attribute(release_last(release), "version", True) != version:
        raise ToolError("unable to find version %s as the most recent release in %s" % (version, path_release))

    if config.build:
        cmd_build(config)

    if config.deploy:
        cmd_deploy(config, dev)

    if config.dist:
        cmd_dist(config, dev)

    log(INFO, "release complete")

    return 0


####################################################################################################################################
def main():
    """Main."""

    # Die silently on SIGPIPE as C programs do, rather than raising when output is piped to a command that exits early
    signal.signal(signal.SIGPIPE, signal.SIG_DFL)

    path_repo = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    # Load the configuration. Logging is not initialized until this succeeds so print any error directly.
    try:
        config = cfg_load(sys.argv[1:], path_repo)
    except ToolError as error:
        print(error)

        return EXIT_ERROR

    log_init(config.log_level, True)
    log(INFO, "%s %s Release Manager" % (PROJECT_NAME, project_version(path_repo)))

    try:
        return command_run(config)
    except KeyboardInterrupt:
        # A ctrl-c is what was asked for, so report it the way the C reports a signal rather than as a stack trace
        log(ERROR, "terminated on signal SIGINT")

        return EXIT_TERM
    except ToolError as error:
        log(ERROR, error)

        return error.status
    except Exception as error:
        # An unexpected exception is a bug here rather than a problem with the release, so show the stack trace
        log(ERROR, error_trace(error, config.log_level >= DEBUG))

        return EXIT_ERROR


if __name__ == "__main__":
    sys.exit(main())
